# Analisa Data Pokemon

In [2]:
import pandas as pd

#### Membuka DataFrame

#### Melihat 5 data pertama dan 5 data terakhir dari dataframe

#        Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1   Bulbasaur  Grass  Poison  318.0  45      49       49       65     65.0   
1  2     Ivysaur  Grass  Poison  405.0  60      62       63       80     80.0   
2  3    Venusaur  Grass  Poison  525.0  80      82       83      100    100.0   
3  4  Charmander   Fire     NaN  309.0  39      52       43       60     50.0   
4  5  Charmeleon   Fire     NaN  405.0  58      64       58       80     65.0   

   Speed  Stage  Legendary  
0     45      1      False  
1     60      2      False  
2     80      3      False  
3     65      1      False  
4     80      2      False

#       Name   Type 1  Type 2  Total   HP  Attack  Defense  Sp. Atk  \
146  147    Dratini   Dragon     NaN  300.0   41      64       45       50   
147  148  Dragonair   Dragon     NaN  420.0   61      84       65       70   
148  149  Dragonite   Dragon  Flying    NaN   91     134       95      100   
149  150     Mewtwo  Psychic     NaN  680.0  106     110       90      154   
150  151        Mew  Psychic     NaN    NaN  100     100      100      100   

     Sp. Def  Speed  Stage  Legendary  
146     50.0     50      1      False  
147     70.0     70      2      False  
148    100.0     80      3      False  
149     90.0    130      1       True  
150    100.0    100      1      False

#### Cek details dari dataframe

(151, 13)


151


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   #          151 non-null    int64  
 1   Name       151 non-null    object 
 2   Type 1     151 non-null    object 
 3   Type 2     67 non-null     object 
 4   Total      146 non-null    float64
 5   HP         151 non-null    int64  
 6   Attack     151 non-null    int64  
 7   Defense    151 non-null    int64  
 8   Sp. Atk    151 non-null    int64  
 9   Sp. Def    143 non-null    float64
 10  Speed      151 non-null    int64  
 11  Stage      151 non-null    int64  
 12  Legendary  151 non-null    bool   
dtypes: bool(1), float64(2), int64(7), object(3)
memory usage: 14.4+ KB
None


Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Stage', 'Legendary'],
      dtype='object')


#### Karena kolumn 'Type 2' diatas tampak anomali (cuman ada 67 non-null, cek column itu)

Type 2
0    Poison
1    Poison
2    Poison
3       NaN
4       NaN
..      ...
146     NaN
147     NaN
148  Flying
149     NaN
150     NaN

[151 rows x 1 columns]

#### Ternyata, index nomor 3 dan 4, value Type 2 nya kosong. cek baris itu

#        Name Type 1 Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
3  4  Charmander   Fire    NaN  309.0  39      52       43       60     50.0   
4  5  Charmeleon   Fire    NaN  405.0  58      64       58       80     65.0   

   Speed  Stage  Legendary  
3     65      1      False  
4     80      2      False

#### Karena Column Sp. Def & Total juga ada yang null, ayo diisi

#### Check Status missing value lagi

In [4]:
df.isnull().sum()

#             0
Name          0
Type 1        0
Type 2       84
Total         5
HP            0
Attack        0
Defense       0
Sp. Atk       0
Sp. Def       8
Speed         0
Stage         0
Legendary     0
dtype: int64

#### Ternyata ada 84 missing value di Type 2. Datanya cuman ada 151, 50% dari data missing. Handlingnya?

#### Final Check

In [5]:
df

#        Name   Type 1  Type 2  Total   HP  Attack  Defense  Sp. Atk  \
0      1   Bulbasaur    Grass  Poison  318.0   45      49       49       65   
1      2     Ivysaur    Grass  Poison  405.0   60      62       63       80   
2      3    Venusaur    Grass  Poison  525.0   80      82       83      100   
3      4  Charmander     Fire     NaN  309.0   39      52       43       60   
4      5  Charmeleon     Fire     NaN  405.0   58      64       58       80   
..   ...         ...      ...     ...    ...  ...     ...      ...      ...   
146  147     Dratini   Dragon     NaN  300.0   41      64       45       50   
147  148   Dragonair   Dragon     NaN  420.0   61      84       65       70   
148  149   Dragonite   Dragon  Flying    NaN   91     134       95      100   
149  150      Mewtwo  Psychic     NaN  680.0  106     110       90      154   
150  151         Mew  Psychic     NaN    NaN  100     100      100      100   

     Sp. Def  Speed  Stage  Legendary  
0       65.0     45      1      False  
1       80.0     60      2      False  
2      100.0     80      3      False  
3       50.0     65      1      False  
4       65.0     80      2      False  
..       ...    ...    ...        ...  
146     50.0     50      1      False  
147     70.0     70      2      False  
148    100.0     80      3      False  
149     90.0    130      1       True  
150    100.0    100      1      False  

[151 rows x 13 columns]

#### Sudah tidak ada kolumn yang ada missing value, mulai analisa

#### Berapa jumlah pokemon yang terdapat di data?,  dan apa saja?

150

array(['Bulbasaur', 'Ivysaur', 'Venusaur', 'Charmander', 'Charmeleon',
       'Charizard', 'Squirtle', 'Wartortle', 'Blastoise', 'Caterpie',
       'Metapod', 'Butterfree', 'Weedle', 'Kakuna', 'Beedrill', 'Pidgey',
       'Pidgeotto', 'Pidgeot', 'Rattata', 'Raticate', 'Spearow', 'Fearow',
       'Ekans', 'Arbok', 'Pikachu', 'Raichu', 'Sandshrew', 'Sandslash',
       'Nidoran���', 'Nidorina', 'Nidoqueen', 'Nidorino', 'Nidoking',
       'Clefairy', 'Clefable', 'Vulpix', 'Ninetales', 'Jigglypuff',
       'Wigglytuff', 'Zubat', 'Golbat', 'Oddish', 'Gloom', 'Vileplume',
       'Paras', 'Parasect', 'Venonat', 'Venomoth', 'Diglett', 'Dugtrio',
       'Meowth', 'Persian', 'Psyduck', 'Golduck', 'Mankey', 'Primeape',
       'Growlithe', 'Arcanine', 'Poliwag', 'Poliwhirl', 'Poliwrath',
       'Abra', 'Kadabra', 'Alakazam', 'Machop', 'Machoke', 'Machamp',
       'Bellsprout', 'Weepinbell', 'Victreebel', 'Tentacool',
       'Tentacruel', 'Geodude', 'Graveler', 'Golem', 'Ponyta', 'Rapidash',
       

#### Pokemon apa yang mempunyai Attack paling tinggi?

#       Name  Type 1  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
148  149  Dragonite  Dragon  405.0  91     134       95      100    100.0   

     Speed  Stage  Legendary  
148     80      3      False

#### Pokemon tipe apa yang mempunyai highest speed?

In [43]:

# cari di df, data yang mempunyai 'Speed' dengan nilai yang sama dengan maximum Speed. Lalu slide agar hanya mendapat 'Type 1'-nya

100    Electric
Name: Type 1, dtype: object

#### Pokemon tipe 1 Dragon punya rata-rata defense berapa?

In [44]:

# cari di df, data yang mempunyai 'Type 1' dengan nilai yang sama dengan 'Dragon'. Lalu slide agar hanya mendapat 'Defense'-nya. Nah, defensenya di cari rata2nya.


68.33333333333333

#### Dari tipe Water or Fire Pokemon, mana yang lebih tinggi rata2 attack speednya?

In [48]:

# cari di df, data yang mempunyai 'Type 1' dengan nilai yang ada di dalam ['Water', 'Fire']. 
# lalu group-in, dan untuk masing2 group, cari rata2nya.
# lalu di slice hanya menampilkan rata2 Sp. Atk

Type 1
Fire     84.583333
Water    64.821429
Name: Sp. Atk, dtype: float64

#### Type Pokemon apa yang lebih banyak terklasifikasi kedalam 'Legendary'?

In [52]:

# cari di df, data yang mempunyai 'Legendary' dengan nilai sama dengan True. 
# lalu group-in, dan untuk masing2 group, cari ada berapa jumlah unique name disana.
# lalu di slice hanya menampilkan jumlah unique nama.

Type 1
Electric    1
Fire        1
Ice         1
Psychic     1
Name: Name, dtype: int64